site/app/libraries/FileUtils.php

```diff
@@ -610,4 +610,91 @@ public static function validateUploadedFiles($files) {
+     * an array of error strings describing the problems.
+     *
+     * Use this function to check that:
+     * - File or directory at $path exists
+     * - File or directory at $path has the correct owner (optional)
+     * - File or directory at $path has the correct group (optional)
+     * - File or directory at $path is readable
+     * - File or directory at $path is writable
+     *
+     * @param string $path Absolute path to file or directory
+     * @param string|null $expected_owner Expected owner name of the file, or null to omit this check
+     * @param string|null $expected_group Expected group name owner of the file, or null to omit this check
+     * @return array Empty array if no errors were detected or an array containing one or more error message strings if
+     *               errors were detected.
+     */
+    public static function checkForPermissionErrors(string $path, ?string $expected_owner, ?string $expected_group): array {
+        $results = [];
+
+        // Check exists
+        $exists = file_exists($path);
+        if (!$exists) {
+            $results[] = "'${path}' does not exist.";
+            return $results;
+        }
+
+        // Check owner
+        if ($expected_owner) {
+            $owner_id = fileowner($path);
+            $owner_name = posix_getpwuid($owner_id)['name'];
+            if ($owner_name !== $expected_owner) {
+                $results[] = "Expected '${path}' to have owner '${expected_owner}' but instead got '${owner_name}'.";
+            }
+        }
+
+        // Check group
+        if ($expected_group) {
+            $group_id = filegroup($path);
+            $group_name = posix_getgrgid($group_id)['name'];
+            if ($group_name !== $expected_group) {
+                $results[] = "Expected '${path}' to have group '${expected_group}' but instead got '${group_name}'.";
+            }
+        }
+
+        // Check is readable
+        $readable = is_readable($path);
+        if (!$readable) {
+            $results[] = "'${path}' is not readable.";
+        }
+
+        // Check is writable
+        $writable = is_writable($path);
+        if (!$writable) {
+            $results[] = "'${path}' is not writable.";
+        }
+
+        return $results;
+    }
+
+    /**
+     * Recursively traverse a directory structure starting at $dir.  The passed in $results array will be populated
+     * with the absolute path to each file or sub-directory.
+     *
+     * NOTE: Sub-directories are treated like files, and thus will be included as their own entries in the array.
+     *
+     * Credit:
+     * https://stackoverflow.com/questions/24783862/list-all-the-files-and-folders-in-a-directory-with-php-recursive-function
+     *
+     * @param string $dir The starting directory (not included in results)
+     * @param array $results An array passed by reference which will be populated
+     */
+    public static function getDirContents(string $dir, &$results = []): void {
```

Comment:
Whats the difference between this function and `getAllFiles` defined in fileUtils ? 
https://github.com/Submitty/Submitty/blob/2e1447add11d77a2d703fd6824a0c8d7bd497bf0/site/app/libraries/FileUtils.php#L31

Reply:
Ignore the other requested changes, i was looking at an older version

Reply:
The result from `getAllFiles()` doesn't include the paths to sub-directories, only to files in those sub-directories.  It also includes a lot of other meta-data I don't need.  I considered using this function but I would have written just as much, if not more code in the caller to parse the result for the useful information I needed.